### PR TITLE
Added actionable errors for network issues.

### LINKED ISSUE #47330: Provide more information on unavailable web app connection
Expected behavior:

Provide more information to the user that a connection failed and how to resolve. 

Current behavior:

If the app access service can't reach a web url the user just gets Bad Gateway. The user wouldn't be clear this is Teleport failing to connect or their own browser.

Image

Bug details:
- Teleport version: 16.4.2
- Recreation steps

1. Add a app service with a url that isn't connected.

```yaml
  - name: "grafana"
    uri: "http://localhost:3000"
    public_addr: "grafana.enterprise.teleportdemo.com"
    labels:
      env: dev
```

2. Attempt to access from the web ui

### PR DESCRIPTION
**Pull request by russjones**
Updated Application Access to return actionable error message when possible for network errors.

Fixes https://github.com/gravitational/teleport/pull/47330

### CODE REVIEW

**Comment by russjones** [comment]
@rosstimothy Adding you to this because I added some timeouts in this PR when I realized no timeouts exists for Application Access.

**Review by stevenGravy — APPROVED** [review]

File: lib/utils/errors.go
Comment (zmb3) [inline lib/utils/errors.go:105]:
```suggestion
		return "Connection refused. Run \"nc -vz HOST PORT\" on the Teleport " +
```

File: lib/utils/errors.go
Comment (zmb3) [inline lib/utils/errors.go:116]:
```suggestion
			"agent to verify a route to the target exists in the " +
```

Consider removing "application" since `lib/utils` is not specific to application access.

File: lib/utils/errors.go
Comment (zmb3) [inline lib/utils/errors.go:124]: Suggesting `curl` only works if the target is an HTTP app. We can get connection reset errors in desktop access or TCP apps (for example), and curl is not going to help there.

File: lib/utils/errors.go
Comment (zmb3) [inline lib/utils/errors.go:133]: The `nc` command here needs a port too, right?

File: lib/utils/errors.go
Comment (zmb3) [inline lib/utils/errors.go:142]: Similarly, if we want these errors to be specific to application access, then I would move this code into `lib/srv/app` or somewhere similar.

File: api/defaults/defaults.go
Comment (zmb3) [inline api/defaults/defaults.go:39]: I'd either move this out of `api/defaults` or name it something like `DeafultAppDialTimeout` since you're only ever using it when dialing apps.

Honestly having a `defaults` package like this was probably never a good idea, it becomes a dumping ground for completely unrelated configuration parameters.

**Review by zmb3 — COMMENTED** [review]
I think this is a slight improvement, but what we've done here is taught the user how to recreate the same error outside of Teleport, not what they need to do to fix it.

For example, today users get confused when they see a connection refused error. After this change merges, they'll see an error that says "run nc -vz.." and when they run that they will see a very similar connection refused error to what they saw with Teleport. This doesn't seem like it will get the user much closer to understanding what's going wrong.

What do you think about making these error messages discuss the cause of the error more? For example, connection refused errors mean that packet(s) did reach the target host, but that nothing was listening on the port, whereas connection timeouts likely indicate some sort of firewall getting in the way.

File: lib/srv/app/transport.go
Comment (russjones) [reply lib/srv/app/transport.go:110]: I could not reproduce this, but I will try again.

**Comment by russjones** [comment]
What do you guys think about the following updated error messages?

```
Connection Refused

Teleport was unable to connect to the requested host, possibly because the server is not
running. Ensure the server is running, listening on the correct port, and not blocked by
firewall rules.

Use "nc -vz HOST PORT" to help debug this issue.
```

```
No Route to Host

Teleport could not connect to the requested host, likely because there is no valid network
path to reach it. Check the network routing table to ensure a valid path to the host exists.

Use "ping HOST" and "ip route get HOST" to help debug this issue.
```

```
Connection Reset by Peer

Teleport could not complete the request because the server abruptly closed the connection
before the response was received. To resolve this issue, ensure the server (or load balancer)
does not have a timeout terminating the connection early and verify that the server is not
crash looping.

Use protocol-specific tools (e.g., curl, psql) to help debug this issue.
```

```
Network I/O Timeout

Teleport was unable to establish a connection to the host within %v, likely due to the
system being overloaded or due to network congestion. To resolve this issue, connect to
the host directly and verify it is responding promptly.

Use protocol-specific tools (e.g., curl, psql) to help debug this issue.
```

```
Context Deadline Exceeded

Teleport did not receive a response within %v, likely due to system overload. To resolve
this issue, connect to the host directly and ensure it is responding promptly.

Use protocol-specific tools (e.g., curl, psql) to assist in debugging this issue.
```

```
No Such Host

Teleport was unable to resolve the provided domain name, likely because the domain does not
exist. To resolve this issue, verify the domain is correct and ensure the DNS resolver is
properly resolving it.

Use "dig +short HOST" to help debug this issue.
```

**Comment by zmb3** [comment]
> Connection Refused
> Teleport was unable to connect to the requested host, possibly because the server is not
running. Ensure the server is running, listening on the correct port, and not blocked by
firewall rules.

I haven't encountered a firewall that results in a connection refused error. Usually they just silently drop packets and the client sees a network timeout.

I'd move firewalls from here to the "context deadline exceeded" error message, otherwise these new messages look good.

File: lib/utils/errors.go
Comment (zmb3) [inline lib/utils/errors.go:135]: Do we want to mention firewalls in this section? (I would)

**Review by zmb3 — APPROVED** [review]

**Review by rosstimothy — APPROVED** [review]

**Review by webvictim — APPROVED** [review]
Thank you @russjones!

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**
